client/allocrunner/task_hook_coordinator.go

```diff
@@ -63,15 +77,29 @@ func (c *taskHookCoordinator) setTasks(tasks []*structs.Task) {
 	if !c.hasPrestartTasks() {
 		c.mainTaskCtxCancel()
 	}
+	if !c.hasMainTasks() {
+		c.poststopTaskCtxCancel()
+	}
+
 }
 
 func (c *taskHookCoordinator) hasPrestartTasks() bool {
-	return len(c.prestartSidecar)+len(c.prestartEphemeral) > 0
+	return len(c.prestartSidecarTasks)+len(c.prestartEphemeralTasks) > 0
+}
+
+func (c *taskHookCoordinator) hasMainTasks() bool {
+	return len(c.mainTasks) > 0
 }
 
 func (c *taskHookCoordinator) startConditionForTask(task *structs.Task) <-chan struct{} {
-	if task.Lifecycle != nil && task.Lifecycle.Hook == structs.TaskLifecycleHookPrestart {
-		return c.closedCh
+	if task.Lifecycle != nil {
+		switch task.Lifecycle.Hook {
+		case structs.TaskLifecycleHookPrestart:
+			// Prestart tasks start without checking status of other tasks
+			return c.closedCh
+		case structs.TaskLifecycleHookPoststop:
+			return c.poststopTaskCtx.Done()
+		}
 	}
 
 	return c.mainTaskCtx.Done()
```

Comment:
You probably can avoid some nesting here - something like:

```go
	if task.Lifecycle == nil {
		return m.mainTaskCtx.Done()
	}

	switch task.Lifecycle.Hook {
	case structs.TaskLifecycleHookPrestart:
		return c.closedCh
	case structs.TaskLifecycleHookPoststop:
		return c.poststopTaskCtx.Done()
	default:
		// unexpected task type - not sure if we should panic here instead
		return c.mainTaskCtx.Done()
	}
```

Reply:
Resolved in 368f894